New implementation of differing module name

gold2718 opened the pull request:
New implementation of differing module name

Implement a solution that allows a Fortran module name to differ from the filename.

User interface changes?: Yes
An optional `module_name` keyword is added to the `ccpp-table-properties` (MetadataTable) section that allows the user to specify a Fortran module name that is independent from the name of the enclosing file.

I feel this PR is a cleaner solution to #636 and should replace #637.

Fixes: #636 

Testing:
  test removed: NA
  unit tests: PASS
  system tests: PASS
  manual testing: manually ran Fortran and unit tests

Commit:
New implementation of differing module name

Commit:
Fix doctests

Commit:
Fix unit test failure caused by previous error message cleanup

climbfuji reviewed (commented):
Apologies if I am not seeing it, but where in the updated tests is the new functionality (scheme name different from file name) tested? I see it for the DDT module name (not equal to file name), though.

Commit:
Add scheme file with different module name as test

gold2718 commented:
> Apologies if I am not seeing it, but where in the updated tests is the new functionality (scheme name different from file name) tested? I see it for the DDT module name (not equal to file name), though.

Good catch! I started with the test mods from #637, however, I just added one (and fixed an error message that I found in the process).

dustinswales commented:
@gold2718 I tested this in the CCPP SCM and there is problem. It seems that if the module has multiple DDTs, it only changes the module_name for the first one in the file (e.g. [radsw_param.f](https://github.com/ufs-community/ccpp-physics/blob/ufs/dev/physics/Radiation/RRTMG/radsw_param.meta)).

Here is a snippet from the Group cap:
```
      use radlw_param,                                  only: sfcflw_type
      use module_radsw_parameters,          only: topfsw_type
      use module_radlw_parameters,           only: topflw_type
      use radsw_param,                                 only: sfcfsw_type
      use GFS_typedefs,                                only: GFS_control_type
      use radsw_param,                                 only: cmpfsw_type
```

Which should be:
```
      use radlw_param,                                  only: sfcflw_type
      use module_radsw_parameters,          only: topfsw_type
      use module_radlw_parameters,           only: topflw_type
      use module_radsw_parameters,          only: sfcfsw_type
      use GFS_typedefs,                                only: GFS_control_type
      use module_radsw_parameters,          only: cmpfsw_type
```

climbfuji reviewed (changes requested):
I am withdrawing my approval for the moment solely based on the conversation in https://github.com/NCAR/ccpp-framework/pull/637#discussion_r1979719579

climbfuji reviewed (approved):
@dustinswales, the other CCPP framework developers, and I discussed that we should fix the CCPP physics so that no host model DDTs need to be passed to the physics instead of updating `capgen`. I did this in https://github.com/ufs-community/ufs-weather-model/pull/2651 and PRs linked in the description. Therefore, I don't see any problems with this PR.

gold2718 commented:
I uploaded a change that allows schemes to use host model DDTs but I now have a conflict in the var_compatibility test that I have to resolve.

Commit:
Allow host model DDTs to be inputs to schemes

peverwhee commented on test/var_compatibility_test/run_test:
```diff
@@ -241,8 +247,8 @@ echo -e "\nChecking lists from command line"
 #check_datatable ${report_prog} ${datafile} "--dependencies" ${dependencies}
-check_datatable ${report_prog} ${datafile} "--suite-list" ${suite_list}       \
-                --sep ";"
+#check_datatable ${report_prog} ${datafile} "--suite-list" ${suite_list}       \
+#                --sep ";"
```
curious as to why this check was commented out?

gold2718 replied:
Because it didn't work? I'm not sure what changed but I fixed it and uncommented those checks.

peverwhee reviewed (approved):
One question, but not enough to stop my approval!

Thanks @gold2718

gold2718 commented on scripts/ccpp_capgen.py:
```diff
@@ -643,9 +643,11 @@ def capgen(run_env, return_db=False):
+        scheme_files = [const_prop_mod] + scheme_files
     # end if
-    scheme_headers, scheme_tdict = parse_scheme_files(scheme_files, run_env)
+    host_ddts = register_ddts(host_files)
```
@dustinswales, life has been a bit busy but I will try to get this all cleaned up in the next few days.

gold2718 replied:
I'm not sure what I was understanding before. I think it should work both ways* but in the version I just pushed**, The host model is using a DDT defined in the scheme list (and not in the host-model list). Please let me know if that looks like the correct test (module_rad_ddt is moved to the suite).

*by "should work both ways", I am referring to how capgen works now, not that I think it is a good idea for schemes to use host-model DDTs

**I'm not sure what I did wrong but I had to force push. It's been a nutty time -- I don't remember rebasing but I gave up trying to fix it.

Commit:
Host model uses DDT defined in schemes.

Commit:
Uncomment and fix var_compatibility_test checks

Commit:
Fix typo

dustinswales reviewed (approved):
@gold2718 Thanks for this cleaner approach!
I've rebased #640 on top of this and all tests pass.

dustinswales commented:
@climbfuji @peverwhee @mkavulich 
We can merge this one after @gold2718 pulls in https://github.com/NCAR/ccpp-framework/pull/657#pullrequestreview-2784573638

climbfuji committed:
Merge branch 'develop' into source_module_ddt_fixes
